snoop/data/analyzers/pdf_preview.py

```diff
@@ -0,0 +1,96 @@
+"""Task to call a service that creates pdf previews for various types of documents.
+
+The service used can be found here: [[https://github.com/thecodingmachine/gotenberg]]
+"""
+
+from .. import models
+from django.conf import settings
+import requests
+from ..tasks import snoop_task, SnoopTaskBroken, returns_json_blob
+import os
+import mimetypes
+
+
+PDF_PREVIEW_MIME_TYPES = {
+    'text/plain',
+    'text/rtf',
+    'application/msword',
+    'vnd.openxmlformats-officedocument.wordprocessingml.document',
+    'application/msexcel',
+    'vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+    'application/mspowerpoint',
+    'application/vnd.openxmlformats-officedocument.presentationml.presentation',
+    'application/vnd.oasis.opendocument.presentation',
+    'application/vnd.oasis.opendocument.spreadsheet',
+    'application/vnd.oasis.opendocument.text',
+}
+"""List of mime types that the pdf generator supports.
+Based on [[https://thecodingmachine.github.io/gotenberg/#office.basic]].
+"""
+
+PDF_PREVIEW_EXTENSIONS = {
+    '.txt',
+    '.rtf',
+    '.fodt',
+    '.doc',
+    '.docx',
+    '.odt',
+    '.xls',
+    '.xlsx',
+    '.ods',
+    '.ppt',
+    '.pptx',
+    '.odp',
+}
+"""List of file extensions that the pdf generator supports.
+Based on [[https://thecodingmachine.github.io/gotenberg/#office.basic]].
+"""
+
+
+def can_create(blob):
+    """Checks if the pdf generator can process this file."""
+    if blob.mime_type in PDF_PREVIEW_MIME_TYPES:
+        return True
+
+
+def call_pdf_generator(data, filename):
+    """Executes HTTP PUT request to the pdf generator service."""
+
+    url = settings.SNOOP_PDF_PREVIEW_URL + 'convert/office'
+
+    resp = requests.post(url, files={'files': (filename, data)})
+
+    if resp.status_code == 504:
+        raise SnoopTaskBroken('pdf generator timed out and returned http 504', 'pdf_preview_http_504')
+
+    if (resp.status_code != 200
+            or resp.headers['Content-Type'] != 'application/pdf'):
+        print(resp.content)
+        raise RuntimeError(f'Unexpected response from pdf generator: {resp}')
+
+    return resp.content
+
+
+@snoop_task('pdf_preview.get_pdf')
+# the @returns_json_blob decorator is only needed to check if this function ran in digests.gather
+@returns_json_blob
+def get_pdf(blob):
+    """Calls the pdf generator for a given blob.
+
+    Adds the pdf preview to the database
+    """
+    filename = models.File.objects.filter(original=blob.pk)[0].name
+    filename_root, ext = os.path.splitext(filename)
+    if ext not in PDF_PREVIEW_EXTENSIONS:
```

Comment:
Same as above: do we really care about extensions at all? Couldn't we replace this with a mime type check? Then, we wouldn't even have an error condition, since we wouldn't be starting the task if the mime types don't match

Reply:
We do need to send a filename with extension, because that's how the service determines which conversion to run. But I replaced the actual filename with a default one. And the check before starting the task and processing the blob is using mime types already, based on the supported extensions.